File: py/RemoveDuplicates.py
def remove_duplicates(nums: list[int]):
    n = len(nums)
    j = 0
    # tmp = ['_'] * n
    if n == 0 or n == 1:
        return n

    for i in range(n - 1):
        if nums[i] != nums[i + 1]:
            nums[j] = nums[i]
            j += 1
    
    # insert the last element
    nums[j] = nums[n - 1]

    # for i in range(j + 1):
    #     nums[i] = tmp[i]

    return j + 1

if __name__ == '__main__':
    # nums = [1, 1, 2]
    nums = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4]

    k = remove_duplicates(nums)
    print(k, nums[:k])

File: cpp/RemoveDuplicates.cpp
# include <iostream>
# include <algorithm>
# include <vector>

using namespace std;

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        int j = 0;
        int n = nums.size();
        
        if (n == 0 || n == 1)
            return n;
        
        for (int i = 0; i < n - 1; i++){
            if (nums[i] != nums[i + 1]){
                nums[j] = nums[i];
                j++;
            }    
        }

        // insert the last element
        nums[j] = nums[n - 1];

        return j + 1;
    }

    int removeDuplicates2(vector<int>& nums){
        sort(nums.begin(), nums.end());
        nums.erase(unique(nums.begin(), nums.end()), nums.end());
        
        int k = nums.size();

        return k;
    }
};

int main(){
    vector<int> nums{ 1, 1, 2, 3, 3, 3, 4, 5, 5};
    
    Solution sol;

    // int k = sol.removeDuplicates(nums);
    // cout << k << "\n";
    // for (int i = 0; i < k; i++){
    //     cout << nums[i] << ", ";
    // }

    int k = sol.removeDuplicates2(nums);
    cout << k << "\n";
    for (auto x: nums)
        cout << x << ", ";

    return 0;
}